sessions/2018/october/Session_15.md

```diff
@@ -0,0 +1,324 @@
+# SOFTWARE ENGINEERING IN JAVA
+
+## Session 15 (16/10/2018)
+
+- Reviewing/fixing Spring API homework
+- API desing concepts
+- Dependency injection concepts
+
+### Notes:
+
+Roger was reviewing my homework solution and advicing me some modifications and improvements.
+
+#### Reviewing/fixing Spring API homework
+
+In certain part of the code I needed to save and incremental `id` `int` field into the object's field that was a `String`. A way to do it should be changing the type of the object's field, but a simpler way (hack) to work around it is doing `id` increment and mapping to `String` by concatenating with `""` at the moment to set into the object's field:
+
+```java
+book.setBookID(""+id++);
+```
+
+__Note__: In the final solution, the object field was changed from `String` to `Integer` (handle `null` values is convenient when answering to request to inexistents elements), since the previous mentioned hack was a workaround.
+
+Another optimization that was spoken was about to change this kind of return conditions when using `Map` structure:
+
+```java
+public Book getBook(String id) {
+    if (bookCache.get(id) != null){
+        return bookCache.get(id);
+    } else {
+        return new Book();
+    }
+}
+```
+
+To use the included `getOrDefault` method that comes integrated into `Map` structure:
+
+```java
+public Book getBook(String id) {
+    return bookCache.getOrDefault(id, new Book());
+    }
+```
+
+#### Use of the API application
+
+When we run the application (see note below), a Tomcat server is started on port `8080`.
+
+__Note__: If changes has been made, the application needs to be restarted. So it's important don't use `Ctrl+Shift+F10` (or the equivalent running by mouse click action), because this will start another Tomcat server instance. Instead of that, what we need to do is go over the bottom panel, and click over the green rerun arrow (first icon on the top of the left vertical menu), or just hit `Ctrl+5` key binding.
+
+__Note__: When we're designing [RESTful][1] interfaces, a main concept is we have to use nouns and not verbs, so `fibonacci_number` is a valid route name, while `calculate` isn't.
+
+##### GET Maps
+
+In order to use the API we have to access GET HTTP routes trought the browser, for example going to <http://localhost:8080/api/customer> we will receive the next HTTP response:
+
+```json
+[
```

Comment:
API design interview question (perhaps, slightly difficult?):
Why is it a bad idea to return an array as the top-level element in an API?
In other words, why is it bad that the first character of an API payload (request or response) is `[` instead of `{`?

(It's OK in our case though - this isn't meant to be a well designed API just yet)

Reply:
Is dangerous because the risk of a [JSON Hijacking Attack](https://haacked.com/archive/2009/06/25/json-hijacking.aspx/), that consists in a GET requests that sends a vulnerable server JSON payload response back to attacker's computer by hijacking the user session cookies, considering that a JSON response contains possible user's sensitive information.

The focal point of this vulnerability comes when the JSON document toplevel denotes an array `[` and not an object `{`, because a script that contains a JSON array is still a valid JavaScript file, so can be executed and used for the attacker purposes.

For example, in the next code, a function is executed each time that an `Ìd` property [is setted][11] into any object, and then returned in a alert (could be sent in back to any location the attacker decides and use for evil purposes):
```html
<html>
...
<body>
    <script type="text/javascript">
        Object.prototype.__defineSetter__('Id', function(obj){alert(obj);});
    </script>
    <script src="http://example.com/Home/AdminBalances"></script>
</body>
</html>
```

Nevertheless, if the response JSON payload begin with an object character `{`, this makes it a non valid JavaScript file and this vulnerability should not work.

So regardless what the JSON format could specify in terms to verify headers or the decision of browser manufactures about how to face script execution, or the choice to only allow POST request in our backend, the best and cheaper (I guess) approach to prevent any possible harm (at least in this attack), should be to use `{` as the JSON top level element in the API response.

__Note__: This seems most popular [browsers already handle this vulnerabilty](https://stackoverflow.com/a/16880162), but I think the [risk is still there](https://portswigger.net/blog/json-hijacking-for-the-modern-web)